Read me update mysql error

tonyxrmdavidson opened the pull request:
This PR will add a Known Issues section to the README.md file. It also adds information on a known issue regarding mysql-db deployment and how to mitigate it when it happens.

## How Has This Been Tested?
Although this PR doesn't have any code changes, the error and workaround it describes has been tested on several openshift clusters using Opendatahub. 

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] The commits and have meaningful messages; the author will squash them [after approval](https://github.com/opendatahub-io/opendatahub-community/blob/main/contributor-cheatsheet.md#:~:text=Usually%20this%20is%20done%20in%20last%20phase%20of%20a%20PR%20revision) or will ask to merge with squash.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

tarilabs commented on test/robot/UserStory.robot:
```diff
@@ -0,0 +1,15 @@
+Resource   UserStory.resource
+
+*** Test Cases ***              
+As a MLOps engineer I would like to store Model name
+    ${name}=  Generate Random String    8    [LETTERS] 
+    ${rId}=  Given I create a RegisteredModel having    name=${name}
+    ${vId}=  And I create a child ModelVersion having    registeredModelID=${rId}    name=v1  
+    ${aId}=  And I create a child ModelArtifact having    modelversionId=${vId}  uri=s3://12345
+    ${r}=  Then I get RegisteredModelByID    id=${rId}
+        And Should be equal    ${r["name"]}    ${name}
```
Demonstrate use of Robot Framework, by implementing (and covering) a User Story from PM going via the REST impl in this PR.

![Screenshot 2023-10-27 at 14 57 23](https://github.com/opendatahub-io/model-registry/assets/1699252/a3aceb73-a6cb-4f48-915e-b7d7497a6630)

Tony, would you deem a sensible use of Robot Framework?
If yes, I'd extend this also beyond "User Story" also to cover for the expected logical mapping of the behind-the-scenes of MLMD (something less of interest of PM and more of us as implementors).

tarilabs reviewed (commented):
Comment for Tony and Radim

tarilabs commented on internal/server/openapi/api_model_registry_service_service.go:
```diff
@@ -115,68 +121,55 @@ func (s *ModelRegistryServiceAPIService) CreateModelArtifact(ctx context.Context
 }
 
 // CreateModelVersionArtifact - Create an Artifact in a ModelVersion
 func (s *ModelRegistryServiceAPIService) CreateModelVersionArtifact(ctx context.Context, modelversionId string, artifact model.Artifact) (ImplResponse, error) {
```
This is something we believe uncovered as we started working on this wirings, and would appreciate your confirmation on our proposed solution below @dhirajsb please?

## Problem statement

In the OpenAPI, _this_ endpoint `CreateModelVersionArtifact` is missing to align to `createModelArtifact`.

Specifically, this endpoint is having as input a generic-but-not-specialized `#/components/schemas/Artifact` 

https://github.com/opendatahub-io/model-registry/blob/52d2151c14eca541ff8187df9df73c2ed60659e6/api/openapi/model-registry.yaml#L417

Missing to provide `artifactType : "ModelArtifact"` in the payload creates a server panic in the code which was autogenerated:

```
net/http.(*conn).serve.func1()
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/server.go:1868 +0xb0
panic({0x102d10cc0?, 0x1033ad3b0?})
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/runtime/panic.go:920 +0x26c
github.com/opendatahub-io/model-registry/internal/server/openapi.AssertArtifactRequired({0x140004b0140?})
	/Users/mmortari/git/model-registry/internal/server/openapi/type_asserts.go:19 +0x5c
github.com/opendatahub-io/model-registry/internal/server/openapi.(*ModelRegistryServiceAPIController).CreateModelVersionArtifact(0x140000ce150, {0x12a33c040, 0x1400020aac0}, 0x140000cc200)
	/Users/mmortari/git/model-registry/internal/server/openapi/api_model_registry_service.go:379 +0x150
net/http.HandlerFunc.ServeHTTP(0x102d09b00?, {0x12a33c040?, 0x1400020aac0?}, 0x140000c0005?)
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/server.go:2136 +0x38
github.com/go-chi/chi/v5.(*Mux).routeHTTP(0x140000a2480, {0x12a33c040, 0x1400020aac0}, 0x140000cc200)
	/Users/mmortari/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.10/mux.go:444 +0x270
net/http.HandlerFunc.ServeHTTP(0x140000d01e0?, {0x12a33c040?, 0x1400020aac0?}, 0x140000cc200?)
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/server.go:2136 +0x38
github.com/go-chi/cors.(*Cors).Handler-fm.(*Cors).Handler.func1({0x12a33c040, 0x1400020aac0}, 0x140000cc200)
	/Users/mmortari/go/pkg/mod/github.com/go-chi/cors@v1.2.1/cors.go:228 +0x19c
net/http.HandlerFunc.ServeHTTP(0x140000cc100?, {0x12a33c040?, 0x1400020aac0?}, 0x102246054?)
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/server.go:2136 +0x38
github.com/go-chi/chi/v5/middleware.init.0.RequestLogger.func1.1({0x102e30cb0, 0x14000324000}, 0x140000cc100)
	/Users/mmortari/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.10/middleware/logger.go:54 +0x120
net/http.HandlerFunc.ServeHTTP(0x102e32818?, {0x102e30cb0?, 0x14000324000?}, 0x1033ae530?)
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/server.go:2136 +0x38
github.com/go-chi/chi/v5.(*Mux).ServeHTTP(0x140000a2480, {0x102e30cb0, 0x14000324000}, 0x140000cc000)
	/Users/mmortari/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.10/mux.go:90 +0x2bc
net/http.serverHandler.ServeHTTP({0x102e2e710?}, {0x102e30cb0?, 0x14000324000?}, 0x6?)
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/server.go:2938 +0xbc
net/http.(*conn).serve(0x14000192090, {0x102e327e0, 0x140000b0db0})
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/server.go:2009 +0x518
created by net/http.(*Server).Serve in goroutine 1
	/opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/server.go:3086 +0x4cc
```

Further, as a result is output again `Artifact` which is not consistent with `createModelArtifact`

## Proposed solution

We propose 2 modification to simplify and make this endpoint consistent:

A. For this endpoint, have as input the `ModelArtifactCreate` as it's the case with `createModelArtifact`:
ie.
make it analogous to:

https://github.com/opendatahub-io/model-registry/blob/52d2151c14eca541ff8187df9df73c2ed60659e6/api/openapi/model-registry.yaml#L73-L80

B. For this endpoint, have as output the `ModelArtifact` as it's the case with `createModelArtifact`:
ie.
make it analogous to:

https://github.com/opendatahub-io/model-registry/blob/52d2151c14eca541ff8187df9df73c2ed60659e6/api/openapi/model-registry.yaml#L83-L85

dhirajsb replied:
Looks like I missed adding the default value for the discriminator property in `ModelArtifact` like so:
```yaml
    ModelArtifact:
      description: An ML model artifact.
      type: object
      allOf:
        -
          $ref: '#/components/schemas/BaseArtifact'
        -
          $ref: '#/components/schemas/ModelArtifactCreate'
        -
          type: object
          properties:
            artifactType:
              type: string
              default: "model-artifact"
```

tarilabs replied:
We applied the suggestion you suggested: https://github.com/opendatahub-io/model-registry/pull/108#discussion_r1375551218
here: 6de0597

We found it did not resolve the panic error detailed https://github.com/opendatahub-io/model-registry/pull/108#discussion_r1374559417 anyway 

At first, we thought it could have been some bug of the openapi generator, as by its doc the case is not explicilty supported: https://github.com/OpenAPITools/openapi-generator/blob/8827da801274a71c88ea2e19538540c6d8c13108/docs/generators/go-server.md?plain=1#L211

But after further analysis, we realized that the methods:
- `Assert{{classname}}Required()`
- `Assert{{classname}}Constraints()`

are out-of-sync even when issuing the Make "gen/openapi-server"
as those are ingnored ( they are generated as part of the "models" )
and seems to us the file `type_asserts.go` is never autogenerated and was possibly a manual-handcrafted file (despite its header).

We are now considering using a custom `model.mustache` template so to have Asserts method regenerated (when manually/explicitly invoking gen/openapi-server) so to have them up-to-date.

tarilabs commented on internal/server/openapi/api_model_registry_service_service.go:
```diff
@@ -115,68 +124,56 @@ func (s *ModelRegistryServiceAPIService) CreateModelArtifact(ctx context.Context
+	// does not even reach the below check which is commented.
+	// panics earlier in type_asserts.go:19
+	// from api_model_registry_service.go:379
+	// if artifact.ModelArtifact == nil {
+	// 	return Response(http.StatusNotImplemented, nil), errors.New("unsupported artifactType")
+	// }
```
Note to self: once Asserts are refreshed, make sure this check is enabled before merging the PR

tarilabs replied:
this was de-commented and part of code now.

tarilabs commented:
@isinyaaa here is a suggestion looking specifically at commit: https://github.com/opendatahub-io/model-registry/pull/108/commits/60756dfc3de2ea0bbb897dc819667c56f167deeb

The [test](https://github.com/opendatahub-io/model-registry/pull/108/commits/60756dfc3de2ea0bbb897dc819667c56f167deeb#diff-e8fee46a378c137ad9fcf43a5bf83180c8f1471b617032f57e48179647f921da) intent is to demonstrate the Logical Model mapping, in this case going REST as input, and [checking on the backend gRPC MLMD](https://github.com/opendatahub-io/model-registry/pull/108/commits/60756dfc3de2ea0bbb897dc819667c56f167deeb#diff-e8fee46a378c137ad9fcf43a5bf83180c8f1471b617032f57e48179647f921daR31-R56).

Kindly consider if you could implement the "robot keywords" in the stubbed file:
https://github.com/opendatahub-io/model-registry/blob/60756dfc3de2ea0bbb897dc819667c56f167deeb/test/robot/MRviaPython.resource#L1-L5
using your client.

When you have some time you could try it out and let us know if that seems to work for you too, please?

 @tonyxrmdavidson and @rkubis : you can reference the commit https://github.com/opendatahub-io/model-registry/pull/108/commits/60756dfc3de2ea0bbb897dc819667c56f167deeb for a concrete demo of what we also shared today.
Do you think it would be possible to parametrize the `.robot` file,
so to execute it once using `MRviaREST.resource` [file](https://github.com/opendatahub-io/model-registry/pull/108/commits/60756dfc3de2ea0bbb897dc819667c56f167deeb#diff-0286f776d2bc501a899733f626ad40cd7c3f63aedb0dc1d05f5c2664ebf8b046),
and a second time using an alternative `.resource` import,
so in the end we would be able to keep 1 "assertion" way but checking going via REST-as-input and Python-as-input, please?

lampajr commented on cmd/proxy.go:
```diff
@@ -25,7 +30,26 @@ hostname and port where it listens.'`,
+		grpc.WithBlock(),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	if err != nil {
+		log.Fatalf("Error dialing connection to mlmd server %s: %v", mlmdAddr, err)
+		return err
+	}
```
We should add this defer to ensure cleaning up the connection resource, wdyt?
```suggestion
	if err != nil {
		log.Fatalf("Error dialing connection to mlmd server %s: %v", mlmdAddr, err)
		return err
	}
	defer conn.Close()
```

lampajr reviewed (approved):
It looks really great. 
Thanks a lot @tarilabs !

Al-Pragliola committed:
fix: made install_protoc.sh compatible with wget2 (#256)

Signed-off-by: Alessio Pragliola <seth.pro@gmail.com>

dependabot[bot] committed:
build(deps): bump aiohttp from 3.10.1 to 3.10.3 in /clients/python (#261)

Bumps [aiohttp](https://github.com/aio-libs/aiohttp) from 3.10.1 to 3.10.3.
- [Release notes](https://github.com/aio-libs/aiohttp/releases)
- [Changelog](https://github.com/aio-libs/aiohttp/blob/master/CHANGES.rst)
- [Commits](https://github.com/aio-libs/aiohttp/compare/v3.10.1...v3.10.3)

---
updated-dependencies:
- dependency-name: aiohttp
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps-dev): bump furo from 2024.7.18 to 2024.8.6 in /clients/python (#262)

Bumps [furo](https://github.com/pradyunsg/furo) from 2024.7.18 to 2024.8.6.
- [Release notes](https://github.com/pradyunsg/furo/releases)
- [Changelog](https://github.com/pradyunsg/furo/blob/main/docs/changelog.md)
- [Commits](https://github.com/pradyunsg/furo/compare/2024.07.18...2024.08.06)

---
updated-dependencies:
- dependency-name: furo
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps-dev): bump mypy from 1.11.0 to 1.11.1 in /clients/python (#263)

Bumps [mypy](https://github.com/python/mypy) from 1.11.0 to 1.11.1.
- [Changelog](https://github.com/python/mypy/blob/master/CHANGELOG.md)
- [Commits](https://github.com/python/mypy/compare/v1.11...v1.11.1)

---
updated-dependencies:
- dependency-name: mypy
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps-dev): bump ruff from 0.5.6 to 0.5.7 in /clients/python (#264)

Bumps [ruff](https://github.com/astral-sh/ruff) from 0.5.6 to 0.5.7.
- [Release notes](https://github.com/astral-sh/ruff/releases)
- [Changelog](https://github.com/astral-sh/ruff/blob/main/CHANGELOG.md)
- [Commits](https://github.com/astral-sh/ruff/compare/0.5.6...0.5.7)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

antisaling committed:
openapi: require name for contexts (#253)

Signed-off-by: Isabella do Amaral <idoamara@redhat.com>

dependabot[bot] committed:
build(deps): bump react-dom from 18.2.0 to 18.3.1 in /clients/ui/frontend (#260)

Bumps [react-dom](https://github.com/facebook/react/tree/HEAD/packages/react-dom) from 18.2.0 to 18.3.1.
- [Release notes](https://github.com/facebook/react/releases)
- [Changelog](https://github.com/facebook/react/blob/main/CHANGELOG.md)
- [Commits](https://github.com/facebook/react/commits/v18.3.1/packages/react-dom)

---
updated-dependencies:
- dependency-name: react-dom
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps-dev): bump style-loader from 3.3.4 to 4.0.0 in /clients/ui/frontend (#258)

Bumps [style-loader](https://github.com/webpack-contrib/style-loader) from 3.3.4 to 4.0.0.
- [Release notes](https://github.com/webpack-contrib/style-loader/releases)
- [Changelog](https://github.com/webpack-contrib/style-loader/blob/master/CHANGELOG.md)
- [Commits](https://github.com/webpack-contrib/style-loader/compare/v3.3.4...v4.0.0)

---
updated-dependencies:
- dependency-name: style-loader
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps-dev): bump rimraf from 5.0.7 to 6.0.1 in /clients/ui/frontend (#259)

Bumps [rimraf](https://github.com/isaacs/rimraf) from 5.0.7 to 6.0.1.
- [Changelog](https://github.com/isaacs/rimraf/blob/main/CHANGELOG.md)
- [Commits](https://github.com/isaacs/rimraf/compare/v5.0.7...v6.0.1)

---
updated-dependencies:
- dependency-name: rimraf
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump @patternfly/react-styles from 6.0.0-alpha.24 to 6.0.0-alpha.34 in /clients/ui/frontend (#257)

Bumps @patternfly/react-styles from 6.0.0-alpha.24 to 6.0.0-alpha.34.

---
updated-dependencies:
- dependency-name: "@patternfly/react-styles"
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

alexcreasy committed:
Adds description and display name properties to mocked ModelRegistry responses (#272)

Signed-off-by: Alex Creasy <alex@creasy.dev>

antisaling committed:
py: update contexts (#270)

Signed-off-by: Isabella do Amaral <idoamara@redhat.com>

antisaling committed:
GHA: build: simplify action (#271)

No need to setup protoc twice. `make build` also invokes `deps`.

Signed-off-by: Isabella do Amaral <idoamara@redhat.com>

dependabot[bot] committed:
build(deps): bump github.com/docker/docker from 27.1.1+incompatible to 27.1.2+incompatible (#274)

Bumps [github.com/docker/docker](https://github.com/docker/docker) from 27.1.1+incompatible to 27.1.2+incompatible.
- [Release notes](https://github.com/docker/docker/releases)
- [Commits](https://github.com/docker/docker/compare/v27.1.1...v27.1.2)

---
updated-dependencies:
- dependency-name: github.com/docker/docker
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): bump aiohttp from 3.10.3 to 3.10.4 in /clients/python (#283)

Bumps [aiohttp](https://github.com/aio-libs/aiohttp) from 3.10.3 to 3.10.4.
- [Release notes](https://github.com/aio-libs/aiohttp/releases)
- [Changelog](https://github.com/aio-libs/aiohttp/blob/master/CHANGES.rst)
- [Commits](https://github.com/aio-libs/aiohttp/compare/v3.10.3...v3.10.4)

---
updated-dependencies:
- dependency-name: aiohttp
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

antisaling committed:
Bump ruff to 0.6.1 (#285)

* build(deps-dev): bump ruff from 0.5.7 to 0.6.1 in /clients/python

Bumps [ruff](https://github.com/astral-sh/ruff) from 0.5.7 to 0.6.1.
- [Release notes](https://github.com/astral-sh/ruff/releases)
- [Changelog](https://github.com/astral-sh/ruff/blob/main/CHANGELOG.md)
- [Commits](https://github.com/astral-sh/ruff/compare/0.5.7...0.6.1)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Fix linter warnings

Signed-off-by: Isabella do Amaral <idoamara@redhat.com>

---------

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Isabella do Amaral <idoamara@redhat.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

tonyxrmdavidson committed:
Adding Known Issues Section
This PR will add a Known Issues section to the README.md file. It also adds informaiton on a known issue regarding DB deployment and how to mitigate it when it happens

Signed-off-by: tonyxrmdavidson <tonyxrmdavidson@yahoo.co.uk>
